File: chapters/chp2_code/curvature.c
#include <stdio.h>
#include <math.h>
#include <stdlib.h>

double polynomial_function(double x) {
    return x*x*x - 3*x*x + x - 2;
}

double first_derivative(double (*func)(double), double x, double dx) {
    return (func(x + dx) - func(x - dx)) / (2 * dx);
}

double second_derivative(double (*func)(double), double x, double dx) {
    return (func(x + dx) - 2 * func(x) + func(x - dx)) / (dx * dx);
}

double curvature(double (*func)(double), double x, double dx) {
    double f_prime = first_derivative(func, x, dx);
    double f_double_prime = second_derivative(func, x, dx);
    double denominator = pow(1 + f_prime * f_prime, 1.5);
    return fabs(f_double_prime) / denominator;
}

void analyze_curvature(double (*func)(double), double x_min, double x_max, 
                      int num_samples, double dx) {
    double max_curvature = 0;
    double max_curvature_x = x_min;
    double step = (x_max - x_min) / (num_samples - 1);
    
    printf("Curvature analysis:\n");
    printf("x\t\tf(x)\t\tκ(x)\n");
    
    for (int i = 0; i < num_samples; ++i) {
        double x = x_min + i * step;
        double fx = func(x);
        double k = curvature(func, x, dx);
        
        printf("%.3f\t\t%.3f\t\t%.6f\n", x, fx, k);
        
        if (k > max_curvature) {
            max_curvature = k;
            max_curvature_x = x;
        }
    }
    
    printf("\nMaximum curvature: %.6f at x = %.3f\n", max_curvature, max_curvature_x);
}

void save_curvature_data(double (*func)(double), double x_min, double x_max,
                        int num_points, const char* filename) {
    FILE* file = fopen(filename, "w");
    if (file == NULL) {
        printf("Error opening file %s\n", filename);
        return;
    }
    
    fprintf(file, "x,y,curvature\n");
    
    double dx = 0.001;
    double step = (x_max - x_min) / (num_points - 1);
    
    for (int i = 0; i < num_points; ++i) {
        double x = x_min + i * step;
        double y = func(x);
        double k = curvature(func, x, dx);
        fprintf(file, "%.6f,%.6f,%.6f\n", x, y, k);
    }
    
    fclose(file);
    printf("Data written to %s\n", filename);
}

int main() {
    printf("Curvature Analysis Tool (C version)\n");
    printf("====================================\n\n");
    
    printf("Example: Polynomial f(x) = x³ - 3x² + x - 2\n");
    
    double dx = 0.001;  // Small increment for numerical differentiation
    
    analyze_curvature(polynomial_function, -1, 4, 11, dx);
    save_curvature_data(polynomial_function, -1, 4, 100, "polynomial_curvature_c.csv");
    
    // Road design example
    printf("\n\nRoad Design Example:\n");
    printf("For a parabolic road y = x², curvature at different points:\n");
    
    for (double t = 0; t <= 5; t += 1) {
        double k = curvature(polynomial_function, t, dx);
        double radius_of_curvature = (k > 0) ? 1.0 / k : INFINITY;
        
        printf("t = %.1f: κ = %.6f, Radius = %.2f", t, k, radius_of_curvature);
        
        if (radius_of_curvature < 50) {
            printf(" (Sharp curve!)");
        }
        printf("\n");
    }
    
    return 0;
}

File: chapters/chp2_code/curvature.rb
#!/usr/bin/env ruby

class CurvatureCalculator
  def initialize(function_proc, dx = 0.001)
    @func = function_proc
    @dx = dx
  end
  
  def first_derivative(x)
    (@func.call(x + @dx) - @func.call(x - @dx)) / (2 * @dx)
  end
  
  def second_derivative(x)
    (@func.call(x + @dx) - 2 * @func.call(x) + @func.call(x - @dx)) / (@dx**2)
  end
  
  def curvature(x)
    f_prime = first_derivative(x)
    f_double_prime = second_derivative(x)
    denominator = (1 + f_prime**2)**1.5
    f_double_prime.abs / denominator
  end
  
  def analyze_curvature(x_min, x_max, num_samples = 10)
    step = (x_max - x_min) / (num_samples - 1).to_f
    max_curvature = 0
    max_curvature_x = x_min
    
    puts "Curvature Analysis:"
    puts "x\t\tf(x)\t\tκ(x)"
    
    (0...num_samples).each do |i|
      x = x_min + i * step
      fx = @func.call(x)
      k = curvature(x)
      
      puts "#{x.round(3)}\t\t#{fx.round(3)}\t\t#{k.round(6)}"
      
      if k > max_curvature
        max_curvature = k
        max_curvature_x = x
      end
    end
    
    puts "\nMaximum curvature: #{max_curvature.round(6)} at x = #{max_curvature_x.round(3)}"
  end
  
  def generate_data(x_min, x_max, num_points, filename, scaling_factor = 1.0)
    step = (x_max - x_min) / (num_points - 1).to_f
    
    File.open(filename, 'w') do |file|
      file.puts "x,y,curvature"
      
      (0...num_points).each do |i|
        x = x_min + i * step
        y = @func.call(x)
        k = curvature(x) * scaling_factor
        file.puts "#{x},#{y},#{k}"
      end
    end
    
    puts "Data written to #{filename}"
  end
end

# Example functions
polynomial = ->(x) { x**3 - 3*x**2 + x - 2 }
sine_wave = ->(x) { Math.sin(x) }
gaussian = ->(x) { Math.exp(-x**2 / 2) }
parabola = ->(x) { x**2 }

puts "Curvature Analysis Tool (Ruby version)"
puts "======================================"
puts

# Example 1: Polynomial function
puts "Example 1: Polynomial f(x) = x³ - 3x² + x - 2"
poly_calc = CurvatureCalculator.new(polynomial)
poly_calc.analyze_curvature(-1, 4, 11)
poly_calc.generate_data(-1, 4, 100, "polynomial_curvature_ruby.csv")

# Example 2: Sine wave
puts "\n\nExample 2: Sine wave f(x) = sin(x)"
sine_calc = CurvatureCalculator.new(sine_wave)
sine_calc.analyze_curvature(-Math::PI, Math::PI, 9)

# Example 3: Gaussian
puts "\n\nExample 3: Gaussian f(x) = exp(-x²/2)"
gauss_calc = CurvatureCalculator.new(gaussian)
gauss_calc.analyze_curvature(-3, 3, 7)

# Road design example
puts "\n\nRoad Design Example: y = x²"
road_calc = CurvatureCalculator.new(parabola)

puts "Curvature at different points along the road:"
(0..5).each do |t|
  k = road_calc.curvature(t)
  radius_of_curvature = k > 0 ? 1.0 / k : Float::INFINITY
  
  print "t = #{t}: κ = #{k.round(6)}, Radius = #{radius_of_curvature.round(2)}"
  
  if radius_of_curvature < 50
    print " (Sharp curve - reduce speed!)"
  end
  puts
end

# Special case analysis
puts "\n\nSpecial Cases:"

# Circle of radius R has constant curvature 1/R
circle_radius = 5
circle = ->(x) { Math.sqrt(circle_radius**2 - x**2) }  # Upper semicircle
circle_calc = CurvatureCalculator.new(circle)

puts "Circle with radius #{circle_radius}:"
puts "Theoretical curvature: #{(1.0/circle_radius).round(6)}"
puts "Calculated curvature at x=0: #{circle_calc.curvature(0).round(6)}"

puts "\nNote: CSV files can be visualized using plotting software."